standards/application/segmentation.md

```diff
@@ -0,0 +1,182 @@
+---
+title: Message Segmentation and Reconstruction over Waku
+name: Message Segmentation and Reconstruction
+tags: [waku-application, segmentation]
+version: 0.1
+status: draft
+---
+
+## Abstract
+
+This specification defines an application-layer protocol for **segmentation** and **reconstruction** of messages carried over Waku when the original payload exceeds the maximum Waku's message size. Applications partition the payload into multiple Waku envelopes and reconstruct the original on receipt, even when segments arrive out of order or up to a **predefined percentage** of segments are lost. The protocol uses **Reed–Solomon** erasure coding for fault tolerance. Messages whose payload size is **≤ `segmentSize`** are sent unmodified.
+
+## Motivation
+
+Waku Relay deployments typically propagate envelopes up to **150 KB** as per [64/WAKU2-NETWORK - Message](https://rfc.vac.dev/waku/standards/core/64/network#message-size). To support larger application payloads, a segmentation layer is required. This specification enables larger messages by partitioning them into multiple envelopes and reconstructing them at the receiver. Erasure-coded parity segments provide resilience against partial loss or reordering.
+
+## Terminology
+
+- **original message**: the full application payload before segmentation.  
+- **data segment**: one of the partitioned chunks of the original message payload.  
+- **parity segment**: an erasure-coded segment derived from the set of data segments.  
+- **segment message**: a Waku envelope whose `payload` field carries a serialized `SegmentMessageProto`.  
+- **`segmentSize`**: configured maximum size in bytes of each data segment’s `payload` chunk (before protobuf serialization).  
+- **sender public key**: the origin identifier used for indexing persistence.
+
+The key words **“MUST”**, **“MUST NOT”**, **“REQUIRED”**, **“SHALL”**, **“SHALL NOT”**, **“SHOULD”**, **“SHOULD NOT”**, **“RECOMMENDED”**, **“NOT RECOMMENDED”**, **“MAY”**, and **“OPTIONAL”** in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+## Segmentation
+
+### Sending
+
+When the original payload exceeds `segmentSize`, the sender **MUST**:
+
+- Compute a 32-byte `entire_message_hash = Keccak256(original_payload)`.
+- Split the payload into one or more **data segments**, each of size up to `segmentSize` bytes.
+- Optionally generate **parity segments** using Reed–Solomon erasure coding, at the predefined parity rate.  
+  Implementations **MUST NOT** produce more than 256 total segments (data + parity).
+- Encode each segment as a `SegmentMessageProto` with:
+  - The `entire_message_hash`
+  - Either data-segment indices (`segments_count`, `index`) or parity-segment indices (`parity_segments_count`, `parity_segment_index`)
```

Comment:
At this point it's not clear what an "index" is (i.e. just a seq number for the segments). We could define inline, but another strategy is to introduce the Wire Protocol (`SegmentMessageProto`) first with a short explanation of the fields, followed by this section that describes how that protobuf is populated. This ordering helps a lot in making sure all terms and concepts are defined before being used in a heuristic.